### DISCUSSION TITLE
Differents on generated pdfs, locally and on production, using dompdf

### DISCUSSION THREAD

**Discussion by aldzic4** [dompdf/dompdf#3250] [Q&A] [answered]
Hello,

I'm generating pdf from php.blade using dompdf, and I'm using custom font FagoPro.
So when I generate it locally, everything looks okay, font, formatted..
But on production it has some mistakes, for example on some parts of html style is not okay.

I will post files, so you can see where is a difference.

Thanks a lot in advantage.
![Screenshot 2023-07-25 232837](https://github.com/dompdf/dompdf/assets/94372357/43cb26f7-919a-4a9c-886a-d9315442edaa)
![Screenshot 2023-07-25 232906](https://github.com/dompdf/dompdf/assets/94372357/0710b98e-39e8-4d8d-96ac-7d60db7c4dde)

**Comment by bsweeney** [comment]
Are you copying your local files to the server? If so, you might need to reset the font directory if the path from the root of the filesystem is not the same.

**Reply by aldzic4** [reply]
But it is interesting that some parts implement font, just only few of them are not okay.

**Reply by bsweeney** [reply]
It is interesting, yes. You can try [enabling additional debug messaging](https://github.com/dompdf/dompdf/wiki/Debugging-Dompdf) to see if you can track down the cause.

**Reply by aldzic4** [reply]
Unfortunately, I couldn't solve it.
But thanks anyway.

**Reply by bsweeney** [reply]
That's too bad. Without more info from your end I can't really provide much more guidance. Post back here if you ever come back to the issue and can provide more details.

**Reply by aldzic4** [reply]
This is my code in laravel, where pdf is generated using dompdf

```
 public function insurancePolicy(Transaction $transaction)
    {

        // dd((json_decode($transaction->purchase_details, true)[2]));
        $tid = $transaction->id;
        $purchaseDetails = json_decode($transaction->purchase_details, true)[2];
        $personalData = json_decode($transaction->purchase_details, true)[3];
        $packageId = $purchaseDetails['package_id'];
        $insuranceId = json_decode($transaction->purchase_details, true)[0]['insurance_type_id'];
        $insurance = InsuranceType::query()->where('id', $insuranceId)->firstOrFail();
        $data = $personalData['personal_data'];
        $package = Package::query()->where('id', $packageId)->first();

        $items = $package->packageItems;
        $transaction = Transaction::query()->where('id', $tid)->first();

        $itemsCoverage = [];
        foreach ($items as $key => $item) {
            $covs = json_decode($item->pivot->coverage);
            $itemsCoverage[] = $covs;
        }

        $pathName = null;

        if (isset($data['first_name'])) {
            $pathName = $data['first_name'] . '-' . $data['last_name'] . '-' . $tid;
        } else {
            $pathName = $data['name'] . '-' . $tid;
        }

        $options = new Options();
        $options->set('defaultFont','FagoPro'); 
        $options->set('isHtml5ParserEnabled', true); 
        $dompdf = new Dompdf($options);

        $dompdf = PDF::loadView('policy/insurance-policy', compact([
            'tid',
            'transaction',
            'package',
            'items',
            'itemsCoverage',
            'insurance'
        ]))
            ->setOption('defaultFont', storage_path('/fonts/FagoPro.otf'));
        $output = $dompdf->output();

        $defaultFont = $dompdf->getOptions()->get('defaultFont');
        // dd('Default Font: ' . $defaultFont);

        $filePath = public_path('tmp/Putno-zdravstveno-osiguranje-' . $pathName . '.pdf');
        file_put_contents($filePath, $output);

        $downloadUrl = asset('tmp/Putno-zdravstveno-osiguranje-' . $pathName . '.pdf');
        // dd($downloadUrl);

        // Return the download response with the PDF file
        return response()->json(['download_url' => $downloadUrl, 'path_name' => $pathName]);
    }
```

**Reply by bsweeney** [reply]
I see you're setting the default font to FagoPro. How have you installed that font into Dompdf?

Also, this line seems problematic:
```
->setOption('defaultFont', storage_path('/fonts/FagoPro.otf'));
```

The DefaultFont option expects a font name that matches an installed font, not the path to the font.

**Reply by aldzic4** [reply]
I did not install font to dompdf, I put .otf file in storage folder, then import it on blade.php file-u, using font-face
And then put to .main CSS class font-family:'FagoPro'

**Reply by bsweeney** [reply]
I think that should work OK. I'd still remove that one line, though Dompdf will fall back to a known font, so ultimately that line doesn't hurt anything.

You might try removing the installed-font.json file from the font directory, if you copied that from your local system.

**Reply by aldzic4** [reply, marked as answer]
Thanks a lot for your help.
I removed installed-font.json file, but also this is how I imported fonts in my function where pdf is generated.

```
$options = new Options();
$options->set('defaultFont', 'FagoPro'); // Set the default font (first font you want to use)
$options->set('isHtml5ParserEnabled', true); // Enable HTML5 parser
$dompdf = new Dompdf($options);
$fontMetrics = new FontMetrics($dompdf->getCanvas(), $options);
$fontMetrics->registerFont(
    [
        'family' => 'FagoPro-Bold',
        'style' => 'normal',
        'weight' => 700,
        'file' => storage_path('fonts/FagoPro-Bold.otf'),
    ],
    storage_path('fonts/FagoPro-Bold.otf')
);
```

This is in blade.php file inside <style> tag:

```
@font-face {
    font-family: 'FagoPro-Bold';
    src: url({{ storage_path("fonts/FagoPro-Bold.otf") }}) format("truetype");
}
```

But, it is important to note that font-weight, eg bold, should not be added to the css class, where you define font-family: custom-font.
Also, the font will not be applied to td tags inside table tags.

**Reply by bsweeney** [reply]
Note that you don't need to install the font programmatically (by calling `registerFont`) as well as via CSS (using `@font-face`). The CSS declaration internally calls registerFont. Without specifying a weight the normal weight is being used so you're registering the same font as both the bold and normal variants.
